compiler/z/codegen/OMRMachine.cpp

```diff
@@ -1927,46 +1782,30 @@ bool
 OMR::Z::Machine::findBestFreeRegisterPair(TR::RealRegister ** firstRegister, TR::RealRegister ** lastRegister,
    TR_RegisterKinds rk, TR::Instruction * currInst, uint64_t availRegMask)
    {
-   uint32_t interference = 0;
-
-   TR::Compilation *comp = self()->cg()->comp();
-
    TR::RealRegister * freeRegisterLow = NULL;
    TR::RealRegister * freeRegisterHigh = NULL;
 
-   uint64_t bestWeightSoFar = (uint64_t) (-1);
-   int32_t iOld = 0, iNew;
+   uint64_t bestWeightSoFar = -1;
 
-   bool enableHighWordRA = self()->cg()->supportsHighWordFacility() && !comp->getOption(TR_DisableHighWordRA) &&
-                           rk != TR_FPR && rk != TR_VRF;
-   bool highWordPairIsFree = true;
-
-   if (rk != TR_FPR  && rk != TR_VRF)
+   if (rk != TR_FPR && rk != TR_VRF)
       {
       // Look at all reg pairs (starting with an even register)
       for (int32_t i = TR::RealRegister::FirstGPR; i <= TR::RealRegister::LastAssignableGPR; i += 2)
          {
          // Don't consider registers that can't be assigned.
-         if ((_registerFile[i + 0]->getState() == TR::RealRegister::Locked) ||
-             (_registerFile[i + 1]->getState() == TR::RealRegister::Locked) ||
-             (_registerFile[i + 0]->getRealRegisterMask() & availRegMask) == 0 ||
-             (_registerFile[i + 1]->getRealRegisterMask() & availRegMask) == 0)
+         if (_registerFile[i + 0]->getState() == TR::RealRegister::Locked || (_registerFile[i + 0]->getRealRegisterMask() & availRegMask) == 0 ||
+             _registerFile[i + 1]->getState() == TR::RealRegister::Locked || (_registerFile[i + 1]->getRealRegisterMask() & availRegMask) == 0)
             {
             continue;
             }
 
-         //self()->cg()->traceRegWeight(_registerFile[i], _registerFile[i]->getWeight());
-         //self()->cg()->traceRegWeight(_registerFile[i + 1], _registerFile[i + 1]->getWeight());
-
-         if (enableHighWordRA && rk == TR_GPR64)
+         // TODO: This is a little too conservative. This API needs to be taught whether the register pair requires
+         // 64-bit registers or not and if and only if we require 64-bit registers do we need to check the highword.
+         bool highWordPairIsFree = false;
+         if ((_registerFile[i + 0]->getHighWordRegister()->getState() == TR::RealRegister::Free || _registerFile[i + 0]->getHighWordRegister()->getState() == TR::RealRegister::Unlatched) &&
+             (_registerFile[i + 1]->getHighWordRegister()->getState() == TR::RealRegister::Free || _registerFile[i + 1]->getHighWordRegister()->getState() == TR::RealRegister::Unlatched))
```

Comment:
Lots of opportunity for improvement. I've left TODOs all around which I'll be addressing after this PR is merged.